crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -1603,18 +1634,44 @@ impl<'src> Parser<'src> {
         }
     }
 
-    pub(super) fn parse_generators(&mut self) -> Vec<ast::Comprehension> {
-        const GENERATOR_SET: TokenSet = TokenSet::new([TokenKind::For, TokenKind::Async]);
-
-        let mut generators = vec![];
-        let mut progress = ParserProgress::default();
+    /// Helper function to validate the comprehension iter and `if` expressions.
+    fn validate_comprehension_iter_and_if_expr(&mut self, parsed_expr: &ParsedExpr) {
+        let ParsedExpr {
+            expr,
+            is_parenthesized,
+        } = parsed_expr;
 
-        while self.at_ts(GENERATOR_SET) {
-            progress.assert_progressing(self);
-            generators.push(self.parse_comprehension());
+        if *is_parenthesized {
+            return;
         }
 
-        generators
+        match expr {
+            Expr::Starred(_) => {
+                self.add_error(
+                    ParseErrorType::OtherError(
+                        "starred expression cannot be used here".to_string(),
+                    ),
+                    expr,
+                );
+            }
```

Comment:
Parenthesized starred expression is also an error but will be reported by `parse_parenthesized_expression` in #10641